### PR TITLE
Better C code + pypy support

### PR DESCRIPTION
**Pull request by asottile**

### CODE REVIEW

**Comment by asottile** [comment]
The interesting bits:

**new**
- py27 benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/66933717
- py34 benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/66933719
- pypy benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/66933720

**old**
- py27 benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/65930964
- py34 benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/65930966
- pypy benchmark: https://travis-ci.org/Yelp/yelp_cheetah/jobs/65930967

Notably:
- locals / globals get slightly slower because they now have an additional assignment and function calls to set up `locals_` and `globals_`, this is acceptable
- searchlist lookups actually *improve*
- pypy kicks the shit outa cpython (now that we're not using C code)

File: Cheetah/legacy_compiler.py [hunk 1]
```diff
@@ -339,7 +339,9 @@ def _addAutoSetupCode(self):
         self.indent()
         self.addChunk('_dummyTrans = False')
         self.dedent()
-        self.addChunk('SL = self._CHEETAH__searchList')
+        self.addChunk('locals_ = locals()')
+        self.addChunk('globals_ = globals()')
```
Comment (bukzor) [inline Cheetah/legacy_compiler.py]: can we do this conditionally, when there's any VFFSL?

File: _cheetah.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    if (PyMapping_HasKeyString(locals, key)) {
+        return PyMapping_GetItemString(locals, key);
+    } else if (PyMapping_HasKeyString(globals, key)) {
+        return PyMapping_GetItemString(globals, key);
+    } else if (PyObject_HasAttrString(_builtins_module, key)) {
+        return PyObject_GetAttrString(_builtins_module, key);
+    } else {
+        return _vfsl(key, selfobj, ns);
+    }
```
Comment (bukzor) [inline _cheetah.c]: this does a hash-lookup twice on the happy paths

File: _cheetah.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+static PyObject* _setup_module(PyObject* module) {
+    if (module) {
+        NotFound = PyErr_NewException("_cheetah.NotFound", PyExc_LookupError, NULL);
+        PyModule_AddObject(module, "NotFound", NotFound);
+
+        _builtins_module = PyImport_ImportModule(IF_PY3("builtins", "__builtin__"));
+        if (!_builtins_module) {
+            Py_DECREF(module);
+            module = NULL;
+        }
```
Comment (bukzor) [inline _cheetah.c]: this seems like a noop where you import builtins then don't use it?

**Comment by bukzor** [comment]
fixnship

File: _cheetah.c (hunk 3 above)
Comment (asottile) [inline _cheetah.c]: it's a static global

File: Cheetah/legacy_compiler.py (hunk 1 above)
Comment (asottile) [inline Cheetah/legacy_compiler.py]: yeah, but we don't have that information at this point.  Can be a followup when we have a better parse tree

File: _cheetah.c (hunk 3 above)
Comment (bukzor) [inline _cheetah.c]: Ok. I see it now.

File: _cheetah.c (hunk 2 above)
Comment (asottile) [inline _cheetah.c]: I guess I could gobble the exception like the implementation of HasAttrString does

File: Cheetah/legacy_compiler.py (hunk 1 above)
Comment (bukzor) [inline Cheetah/legacy_compiler.py]: inline a TODO plz

**Pull request merged**